File: concepts/type_system/typescript/type_system.ts
// 型システム - TypeScript 実装
//
// TypeScript は構造的部分型 (Structural Subtyping) を採用
// 名前ではなく構造で型の互換性を判定

console.log("=== Type System Demo ===\n");

// --- 基本型 ---
function basicTypes(): void {
  console.log("--- 基本型 ---");

  // プリミティブ型
  const str: string = "Hello, TypeScript!";
  const num: number = 42;
  const float: number = 3.14; // number は整数も小数も
  const bool: boolean = true;
  const n: null = null;
  const u: undefined = undefined;

  console.log(`string: ${str}`);
  console.log(`number (int): ${num}`);
  console.log(`number (float): ${float}`);
  console.log(`boolean: ${bool}`);
  console.log(`null: ${n}, undefined: ${u}`);

  // bigint
  const big: bigint = 9007199254740991n;
  console.log(`bigint: ${big}`);

  // symbol
  const sym: symbol = Symbol("unique");
  console.log(`symbol: ${sym.toString()}`);

  console.log();
}

// --- 配列とタプル ---
function arraysAndTuples(): void {
  console.log("--- 配列とタプル ---");

  // 配列
  const numbers: number[] = [1, 2, 3, 4, 5];
  const strings: Array<string> = ["a", "b", "c"];
  console.log(`number[]: ${numbers}`);
  console.log(`Array<string>: ${strings}`);

  // タプル (固定長・各要素の型が異なる)
  const tuple: [string, number, boolean] = ["hello", 42, true];
  console.log(`tuple: ${tuple}`);

  // 名前付きタプル
  type Point = [x: number, y: number];
  const point: Point = [10, 20];
  console.log(`named tuple: (${point[0]}, ${point[1]})`);

  // 可変長タプル
  type StringNumberPairs = [string, ...number[]];
  const pairs: StringNumberPairs = ["values", 1, 2, 3];
  console.log(`variadic tuple: ${pairs}`);

  console.log();
}

// --- オブジェクト型 ---
function objectTypes(): void {
  console.log("--- オブジェクト型 ---");

  // インラインオブジェクト型
  const user: { name: string; age: number } = {
    name: "Alice",
    age: 30,
  };
  console.log(`object: ${JSON.stringify(user)}`);

  // インターフェース
  interface Person {
    name: string;
    age: number;
    email?: string; // オプショナル
    readonly id: number; // 読み取り専用
  }

  const person: Person = { name: "Bob", age: 25, id: 1 };
  console.log(`interface: ${JSON.stringify(person)}`);

  // 型エイリアス
  type Coordinate = {
    x: number;
    y: number;
  };

  const coord: Coordinate = { x: 10, y: 20 };
  console.log(`type alias: ${JSON.stringify(coord)}`);

  console.log();
}

// --- ユニオン型と交差型 ---
function unionAndIntersection(): void {
  console.log("--- ユニオン型と交差型 ---");

  // ユニオン型 (OR)
  type StringOrNumber = string | number;
  const val1: StringOrNumber = "hello";
  const val2: StringOrNumber = 42;
  console.log(`union: ${val1}, ${val2}`);

  // リテラル型とユニオン
  type Direction = "north" | "south" | "east" | "west";
  const dir: Direction = "north";
  console.log(`literal union: ${dir}`);

  // 交差型 (AND)
  type Named = { name: string };
  type Aged = { age: number };
  type NamedAndAged = Named & Aged;

  const person: NamedAndAged = { name: "Alice", age: 30 };
  console.log(`intersection: ${JSON.stringify(person)}`);

  // Discriminated Union (判別可能なユニオン)
  type Shape =
    | { kind: "circle"; radius: number }
    | { kind: "rectangle"; width: number; height: number };

  function area(shape: Shape): number {
    switch (shape.kind) {
      case "circle":
        return Math.PI * shape.radius ** 2;
      case "rectangle":
        return shape.width * shape.height;
    }
  }

  console.log(`discriminated union: area = ${area({ kind: "circle", radius: 5 }).toFixed(2)}`);

  console.log();
}

// --- 型推論 ---
function typeInference(): void {
  console.log("--- 型推論 ---");

  // 型推論
  const x = 42; // number と推論
  const y = "hello"; // string と推論
  const z = [1, 2, 3]; // number[] と推論

  console.log(`x: number = ${x}`);
  console.log(`y: string = ${y}`);
  console.log(`z: number[] = ${z}`);

  // const アサーション
  const config = {
    endpoint: "/api",
    timeout: 3000,
  } as const;
  // 型: { readonly endpoint: "/api"; readonly timeout: 3000 }

  console.log(`as const: ${JSON.stringify(config)}`);

  // 配列の const アサーション
  const directions = ["north", "south"] as const;
  // 型: readonly ["north", "south"]
  console.log(`const array: ${directions}`);

  console.log();
}

// --- ジェネリクス ---
function generics(): void {
  console.log("--- ジェネリクス ---");

  // ジェネリック関数
  function identity<T>(value: T): T {
    return value;
  }

  console.log(`identity<number>: ${identity(42)}`);
  console.log(`identity<string>: ${identity("hello")}`);

  // 制約付きジェネリクス
  function getLength<T extends { length: number }>(item: T): number {
    return item.length;
  }

  console.log(`getLength("hello"): ${getLength("hello")}`);
  console.log(`getLength([1,2,3]): ${getLength([1, 2, 3])}`);

  // ジェネリッククラス
  class Box<T> {
    constructor(private value: T) {}
    get(): T {
      return this.value;
    }
  }

  const numBox = new Box(42);
  const strBox = new Box("hello");
  console.log(`Box<number>: ${numBox.get()}`);
  console.log(`Box<string>: ${strBox.get()}`);

  // 複数の型パラメータ
  function map<T, U>(arr: T[], fn: (item: T) => U): U[] {
    return arr.map(fn);
  }

  const doubled = map([1, 2, 3], (x) => x * 2);
  console.log(`map: ${doubled}`);

  console.log();
}

// --- 型ガードとナローイング ---
function typeGuards(): void {
  console.log("--- 型ガードとナローイング ---");

  // typeof
  function process(value: string | number): string {
    if (typeof value === "string") {
      return value.toUpperCase(); // string と推論
    }
    return value.toFixed(2); // number と推論
  }

  console.log(`typeof guard (string): ${process("hello")}`);
  console.log(`typeof guard (number): ${process(3.14159)}`);

  // instanceof
  class Dog {
    bark() {
      return "Woof!";
    }
  }
  class Cat {
    meow() {
      return "Meow!";
    }
  }

  function speak(animal: Dog | Cat): string {
    if (animal instanceof Dog) {
      return animal.bark();
    }
    return animal.meow();
  }

  console.log(`instanceof guard: ${speak(new Dog())}`);

  // in 演算子
  type Fish = { swim: () => void };
  type Bird = { fly: () => void };

  function move(animal: Fish | Bird): string {
    if ("swim" in animal) {
      return "swimming";
    }
    return "flying";
  }

  console.log(`in guard: ${move({ swim: () => {} })}`);

  // カスタム型ガード
  function isString(value: unknown): value is string {
    return typeof value === "string";
  }

  const val: unknown = "hello";
  if (isString(val)) {
    console.log(`custom guard: ${val.toUpperCase()}`);
  }

  console.log();
}

// --- ユーティリティ型 ---
function utilityTypes(): void {
  console.log("--- ユーティリティ型 ---");

  interface User {
    name: string;
    age: number;
    email: string;
  }

  // Partial
  type PartialUser = Partial<User>;
  const partial: PartialUser = { name: "Alice" };
  console.log(`Partial<User>: ${JSON.stringify(partial)}`);

  // Required
  type RequiredUser = Required<PartialUser>;

  // Pick
  type UserName = Pick<User, "name" | "age">;
  const picked: UserName = { name: "Bob", age: 25 };
  console.log(`Pick<User, "name" | "age">: ${JSON.stringify(picked)}`);

  // Omit
  type UserWithoutEmail = Omit<User, "email">;
  const omitted: UserWithoutEmail = { name: "Charlie", age: 30 };
  console.log(`Omit<User, "email">: ${JSON.stringify(omitted)}`);

  // Record
  type UserMap = Record<string, number>;
  const scores: UserMap = { Alice: 100, Bob: 85 };
  console.log(`Record<string, number>: ${JSON.stringify(scores)}`);

  // Readonly
  type ReadonlyUser = Readonly<User>;

  // ReturnType
  function getUser(): User {
    return { name: "Alice", age: 30, email: "alice@example.com" };
  }
  type GetUserReturn = ReturnType<typeof getUser>;
  console.log(`ReturnType: User`);

  console.log();
}

// --- 実行 ---
basicTypes();
arraysAndTuples();
objectTypes();
unionAndIntersection();
typeInference();
generics();
typeGuards();
utilityTypes();

File: concepts/type_system/go/type_system.go
// 型システム - Go 実装
//
// Go は静的型付け言語
// 型推論、インターフェース、ジェネリクス (1.18+) をサポート

package main

import (
	"fmt"
)

func main() {
	fmt.Println("=== Type System Demo ===\n")

	basicTypes()
	compositeTypes()
	typeInference()
	interfaces()
	generics()
	typeAssertions()
}

// --- 基本型 ---
func basicTypes() {
	fmt.Println("--- 基本型 ---")

	// 整数型
	var i int = 42
	var i8 int8 = 127
	var i64 int64 = 9223372036854775807
	fmt.Printf("int: %d, int8: %d, int64: %d\n", i, i8, i64)

	// 符号なし整数
	var u uint = 42
	var u8 uint8 = 255 // byte
	fmt.Printf("uint: %d, uint8: %d\n", u, u8)

	// 浮動小数点
	var f32 float32 = 3.14
	var f64 float64 = 3.141592653589793
	fmt.Printf("float32: %f, float64: %f\n", f32, f64)

	// 真偽値
	var b bool = true
	fmt.Printf("bool: %t\n", b)

	// 文字列
	var s string = "Hello, Go!"
	fmt.Printf("string: %s\n", s)

	// rune (Unicode コードポイント)
	var r rune = '日'
	fmt.Printf("rune: %c (U+%04X)\n", r, r)

	fmt.Println()
}

// --- 複合型 ---
func compositeTypes() {
	fmt.Println("--- 複合型 ---")

	// 配列 (固定長)
	var arr [3]int = [3]int{1, 2, 3}
	fmt.Printf("Array: %v (len=%d)\n", arr, len(arr))

	// スライス (可変長)
	slice := []int{1, 2, 3, 4, 5}
	slice = append(slice, 6)
	fmt.Printf("Slice: %v (len=%d, cap=%d)\n", slice, len(slice), cap(slice))

	// マップ
	m := map[string]int{
		"one":   1,
		"two":   2,
		"three": 3,
	}
	fmt.Printf("Map: %v\n", m)

	// 構造体
	type Point struct {
		X, Y int
	}
	p := Point{X: 10, Y: 20}
	fmt.Printf("Struct: %+v\n", p)

	// ポインタ
	ptr := &p
	fmt.Printf("Pointer: %p, *ptr = %+v\n", ptr, *ptr)

	fmt.Println()
}

// --- 型推論 ---
func typeInference() {
	fmt.Println("--- 型推論 ---")

	// := で型推論
	x := 42         // int
	y := 3.14       // float64
	z := "hello"    // string
	b := true       // bool

	fmt.Printf("x: %T = %v\n", x, x)
	fmt.Printf("y: %T = %v\n", y, y)
	fmt.Printf("z: %T = %v\n", z, z)
	fmt.Printf("b: %T = %v\n", b, b)

	// 複合リテラルの型推論
	slice := []int{1, 2, 3}
	m := map[string]int{"a": 1}
	fmt.Printf("slice: %T\n", slice)
	fmt.Printf("map: %T\n", m)

	fmt.Println()
}

// --- インターフェース ---
func interfaces() {
	fmt.Println("--- インターフェース ---")

	// インターフェース定義
	type Shape interface {
		Area() float64
	}

	// 構造体
	type Rectangle struct {
		Width, Height float64
	}

	type Circle struct {
		Radius float64
	}

	// メソッド実装 (暗黙的にインターフェースを満たす)
	// Rectangle に Area メソッドを追加
	// (この例ではローカル関数内なので実際には追加できないが、概念を示す)

	// 代わりに既存の型でデモ
	fmt.Println("インターフェースは暗黙的に実装される")
	fmt.Println("implements キーワードは不要")

	// Stringer インターフェース
	type Person struct {
		Name string
		Age  int
	}
	p := Person{Name: "Alice", Age: 30}
	fmt.Printf("Person: %+v\n", p)

	// 空インターフェース (any)
	var anything interface{} = 42
	fmt.Printf("any: %v (type: %T)\n", anything, anything)
	anything = "hello"
	fmt.Printf("any: %v (type: %T)\n", anything, anything)

	fmt.Println()
}

// --- ジェネリクス (Go 1.18+) ---
func generics() {
	fmt.Println("--- ジェネリクス (Go 1.18+) ---")

	// ジェネリック関数
	result := Min(3, 5)
	fmt.Printf("Min(3, 5) = %d\n", result)

	resultF := Min(3.14, 2.71)
	fmt.Printf("Min(3.14, 2.71) = %f\n", resultF)

	// ジェネリックスライス操作
	ints := []int{1, 2, 3}
	strings := []string{"a", "b", "c"}
	fmt.Printf("Contains(ints, 2) = %t\n", Contains(ints, 2))
	fmt.Printf("Contains(strings, \"x\") = %t\n", Contains(strings, "x"))

	// ジェネリックスタック
	stack := &Stack[int]{}
	stack.Push(1)
	stack.Push(2)
	stack.Push(3)
	fmt.Printf("Stack: %v\n", stack.items)

	val, ok := stack.Pop()
	fmt.Printf("Pop: %d (ok=%t)\n", val, ok)

	fmt.Println()
}

// ジェネリック Min 関数
type Ordered interface {
	~int | ~int8 | ~int16 | ~int32 | ~int64 |
		~uint | ~uint8 | ~uint16 | ~uint32 | ~uint64 |
		~float32 | ~float64 | ~string
}

func Min[T Ordered](a, b T) T {
	if a < b {
		return a
	}
	return b
}

// ジェネリック Contains 関数
func Contains[T comparable](slice []T, value T) bool {
	for _, v := range slice {
		if v == value {
			return true
		}
	}
	return false
}

// ジェネリック Stack
type Stack[T any] struct {
	items []T
}

func (s *Stack[T]) Push(item T) {
	s.items = append(s.items, item)
}

func (s *Stack[T]) Pop() (T, bool) {
	if len(s.items) == 0 {
		var zero T
		return zero, false
	}
	item := s.items[len(s.items)-1]
	s.items = s.items[:len(s.items)-1]
	return item, true
}

// --- 型アサーション ---
func typeAssertions() {
	fmt.Println("--- 型アサーション ---")

	var i interface{} = "hello"

	// 型アサーション
	s, ok := i.(string)
	fmt.Printf("i.(string) = %q, ok = %t\n", s, ok)

	n, ok := i.(int)
	fmt.Printf("i.(int) = %d, ok = %t\n", n, ok)

	// 型スイッチ
	checkType := func(v interface{}) string {
		switch v := v.(type) {
		case int:
			return fmt.Sprintf("int: %d", v)
		case string:
			return fmt.Sprintf("string: %s", v)
		case bool:
			return fmt.Sprintf("bool: %t", v)
		default:
			return fmt.Sprintf("unknown: %T", v)
		}
	}

	fmt.Println(checkType(42))
	fmt.Println(checkType("hello"))
	fmt.Println(checkType(true))
	fmt.Println(checkType(3.14))
}
